flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,590 @@
+        // TODO: see FLINK-23154
+        Assume.assumeTrue(
+                "This tests only succeeds if " + ConfigConstants.ENV_FLINK_OPT_DIR + " is set.",
+                System.getProperty(ConfigConstants.ENV_FLINK_OPT_DIR) != null);
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null,
+                        Collections.emptyList(),
+                        null,
+                        PackagedProgramUtils.getPythonDriverClassName(),
+                        new String[0]);
+        assertThat(
+                informationProvider.getJobClassName(),
+                is(PackagedProgramUtils.getPythonDriverClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJarFile().get(),
+                is(PackagedProgramUtils.getPythonJar().getFile()));
+    }
+
+    @Test
+    public void deriveEntryClassInformationForPythonBasedOnParameter() throws FlinkException {
+        // TODO: see FLINK-23154
+        Assume.assumeTrue(
+                "This tests only succeeds if " + ConfigConstants.ENV_FLINK_OPT_DIR + " is set.",
+                System.getProperty(ConfigConstants.ENV_FLINK_OPT_DIR) != null);
+
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null, Collections.emptyList(), null, null, new String[] {"--python"});
+
+        assertThat(
+                informationProvider.getJobClassName(),
+                is(PackagedProgramUtils.getPythonDriverClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJarFile().get(),
+                is(PackagedProgramUtils.getPythonJar().getFile()));
+    }
+
+    @Test
+    public void deriveEntryClassInformationForCustomJar()
+            throws FlinkException, MalformedURLException {
+        // make loading from system classpath fail to make sure that it's not triggered
+        classpathProvider.setSystemClasspathWithTwoEntryClasses();
+
+        final String jobClassName = "SomeJobClassName";
+        final File jarFile = new File("some/jar/file.jar");
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null, null, jarFile, jobClassName, new String[0]);
+        assertThat(informationProvider.getJobClassName(), is(jobClassName));
+        assertThat(informationProvider.getJarFile().isPresent(), is(true));
+        assertThat(informationProvider.getJarFile().get(), is(jarFile));
+    }
+
+    @Test
+    // TODO: we might want to change this behavior triggering a failure here
```

Comment:
Is this a right concern?

Reply:
I don't understand the comment, please elaborate.

Reply:
Right now, we can specify a class and don't check whether it's actual present on the classpath. I was wondering whether that's something we want to change. We're already checking whether the class is available in the user classpath. We could do the same for the system classpath.